### ISSUE TITLE
IPv6 default route for replies only

### ISSUE THREAD

**Issue by grawity** [zerotier/ZeroTierOne#372]
I've successfully made ZeroTier assign global IPv6 addresses, and I want the hosts to be reachable via IPv6 from everywhere, _but_ I don't want ZeroTier to become the default route for everything (due to considerably higher latency).

On Linux, I can do that by creating a separate routing table, and a policy rule based on the "from" address:

```
ip -6 route add ::/0 via fd80:56c2:…/88 dev zt1 table 9993
ip -6 rule add from 2001:db8:…/88 lookup 9993
```

Would it be possible for ZeroTier to do this automatically, or should I stick with shellscripts? And if it's the latter, can the ZeroTier daemon at least run a script once it has configured the network? (I _really_ don't want to do race-condition-y things like "start zerotier; sleep 15; add routes".)

**Comment by adamierymenko** [comment]
What happens if you turn off `allowDefault` on your local ZT instance but leave `allowGlobal`? In that case ZT will still accept and assign the IP but will not override the default route.

What happens then? Does Linux send replies out via the interface they came in on? Or does it try to use the default route, resulting in an incomplete circuit?

**Comment by grawity** [comment]
It's the latter – Linux always sends replies according to the routing table; so if the host has native IPv6, most ISPs will block the 'wrong' source addresses. (My laptops usually don't have v6 _at all_, though, so the replies aren't sent anywhere.)

**Comment by adamierymenko** [comment]
Nearly all ISPs will block the wrong source address, which is why "IPv6 mobility" is not going anywhere (pun intended?).

Going to make this into a backlog entry since this is interesting.

BTW on Windows I think it *does* treat the device as 'sticky'. Not sure what Mac does.

**Comment by adamierymenko** [comment]
This basically gives you a mobile IPv6 address without messing with your default routes. Can see lots of use cases for that.

**Comment by adamierymenko** [comment]
I wonder if this should be default behavior if `allowGlobal` is on but `allowDefault` is not?

**Comment by adamierymenko** [comment]
Or if no default route is set.

**Comment by adamierymenko** [comment]
Personally I think this is the "expected" behavior and Linux's behavior is unexpected and counter-intuitive. Windows actually does the right thing here. Might be the first time I have ever found a case where Windows does the right thing networking-wise.

**Comment by grawity** [comment]
> Nearly all ISPs will block the wrong source address, which is why "IPv6 mobility" is not going anywhere (pun intended?).

Ah, that would have been _IPv4_ mobility; I'm sure I've read that Mobile IPv6 ditched the triangular routing precisely because of that reason. (Though during my tests with ZeroTier, when pinging the laptop from various servers, I did end up receiving replies _with the fd80:: address as source_...)

> This basically gives you a mobile IPv6 address without messing with your default routes. Can see lots of use cases for that.

Right. My current approach doesn't actually give me _outgoing_ IPv6 connectivity like Teredo would (again, I prefer native v4 over high-latency tunneled v6), but it does give me a fixed address for _incoming_ connections, which I find very useful.

~

Alternatively, it might _probably_ be enough if `allowDefault` could be configured to add just `::/0` with high metric (i.e. low priority), rather than trying to override everything. That way, when there's native v6, it would take priority over ZeroTier. And when there _isn't_ native v6, then /etc/gai.conf could be configured to prefer native v4 over ZeroTier-tunneled v6 (that's what Teredo also does).

~

Well, anyway, in case this feature request ends up more work than expected, then at least ability to run "post-up" hook scripts would be also useful.

**Comment by adamierymenko** [comment]
Would it work to just make the metric configurable then?

**Comment by adamierymenko** [comment]
That way if you had no IPv6 connectivity your traffic would go via ZT, which would mean you would always have *some* IPv6.

**Comment by grawity** [comment]
Hmm, now that I think of it, I guess the `ip rule` thing is _still_ required for proper multihoming. That is, I'd still want to ssh into the ZeroTier address no matter what other connectivity the laptop has – and if it had native IPv6, then I'd still have the problem with replies going out the wrong interface. So nevermind that option, I guess...

(Actually I've noticed the metric & flags parameters in the config API, but haven't tried them yet, because the split `::/1 + 8000::/1` routes that ZeroTier adds would have overridden everything else, making metric irrelevant.)

**Comment by adamierymenko** [comment]
Metrics and flags in the config API are currently unused.

**Comment by grawity** [comment]
Playing around with this a bit more, looks like it doesn't work entirely the way I imagined. That is, `ip -6 rule` is certainly useful for the default route selection – I don't know why Linux hasn't made this automatic yet (strong host model). But the other part of my plan (incoming IPv6 connectivity but not outgoing) appears to have been mostly accidental.

It just happened that connect() always chose the local address from another interface (VM bridge) with a lower ifindex, which didn't match the rule and therefore didn't have a default route. But if zt0 were the first (or only) interface with an IPv6 address, it'd still be used by default.

Instead, to prefer native v6 → native v4 → ZeroTier v6, editing of `/etc/gai.conf` is required (that's how Teredo is ranked down as well; Windows has the same under `netsh ipv6 prefixpolicy`), and that's probably something best left to the system admin.

In conclusion, adding an `ip rule` would still be useful, but not as much as I'd thought. (I'd still like that post-up shellscript hook though.)

**Comment by grawity** [comment]
Also, thinking even more about it, the rule I've added also ends up causing _all other routes_ in the `main` table to be ignored as well... (For example, if I add `So even more policy routing is necessary, I imagine it'd have to be like this:

  1. `from all lookup local` (stock default rule)
  1. `from all lookup main suppress_prefixlength 1` (copy of bottom-most rule, but with ::/0 suppressed)
  2. `from 2001:470::.../88 lookup default_via_zt1` (has ::/0 for zt1)
  3. `from 2001:778::…/64 lookup default_via_eth0` (has ::/0 for eth0)
  3. `from all lookup main` (stock default rule)

In the end, it looks like this is not something ZeroTier could properly do on its own. Pity.

**Comment by grawity** [comment]
So I found a way to do this without screwing around with `ip rule`. Turns out Linux actually supports source-based routing under `CONFIG_IPV6_SUBTREES`. So I can just

```
ip -6 route add ::/0 from 2001:db8:dd9f:…:9300::/88 via fd80:56c2:e21c:…:9347:19b3:7753 dev zt1
ip -6 route add ::/0 from 2002:c0a8:3c3:1::/64 via fe80::6e3b:feff:6bfc:3b9c dev wlan0
```

and it works without much trouble. So ZeroTier could do the same when installing a default route.

**Comment by adamierymenko** [comment]
so you just add the 'via'?

**Comment by grawity** [comment]
No, `via` is part of the regular syntax; it's `from <source_prefix>` that's new &ndash; as opposed to just:

```
ip -6 route add ::/0 via fe80::6e3b:feff:6bfc:3b9c dev wlan0
```

This appears to correspond to RTA_SRC in netlink.

**Comment by adamierymenko** [comment]
ahh -- I see. Will investigate this though probably won't hold up 1.2.0 for it.

**Comment by grawity** [comment]
Sure, it's mostly just a "nice to have" feature for later.

Forgot to mention: according to babeld (another software using this feature), ipv6_subtrees is "usable" as of kernel 3.11; while it seems to have existed in older versions it was useless before that.

**Comment by adamierymenko** [comment]
so setting it would be harmless

**Comment by rpodgorny** [comment]
any idea how to achieve the same thing (no extra routing table) for ipv4?

**Comment by grawity** [comment]
@rpodgorny There's no other method besides policy rules, afaik, as all work on multihoming seems to be focused on IPv6.
